CompilerSource/parsing/macros.cpp

```diff
@@ -0,0 +1,316 @@
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "macros.h"
+
+#include <cstring>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <vector>
+using namespace std;
+
+#include "macros.h"
+#include "lexer.h"
+using namespace jdi;
+
+using std::string;
+
+namespace enigma {
+namespace parsing {
+namespace {
+
+string QuoteString(std::string_view str) {
+  std::string res;
+  res.reserve(str.length() * 2 + 2);
+  res += '"';
+  for (char c : str) switch (c) {
+    case '\r': res += "\\r"; break;
+    case '\n': res += "\\n"; break;
+    case '\\': res += "\\\\"; break;
+    case '\"': res += "\\\""; break;
+    case '\'': res += "\\\'"; break;
+    default: res += c;
+  }
+  res += '"';
+  return res;
+}
+
+}  // namespace
+
+string Macro::NameAndPrototype() const {
+  string res = name;
+  if (parameters) {
+    res += "(";
+    for (size_t i = 0; i < parameters->size(); i++)
+      res += (*parameters)[i]
+          + (i + 1 < parameters->size() ? ", " : is_variadic ? "..." : "");
+    res += ")";
+  }
+  return res;
+}
+
+string Macro::ToString() const {
+  string res = "#define " + NameAndPrototype() + " \\\n";
+  for (size_t i = 0; i < value.size(); ++i)
+    res += "  " + value[i].ToString() + (i + 1 < value.size()? "\\\n" : "");
+  return res;
+}
+
+constexpr char kConcatenationError[] =
+    "Concatenation marker cannot appear at either end of a replacement list.";
+
+inline bool meaningful_span(const TokenVector &tokens, size_t b, size_t e) {
+  for (auto i = b; i < e; ++i) {
+    if (!tokens[i].PreprocessesAway()) return true;
+  }
+  return false;
+}
+
+vector<Macro::FuncComponent> Macro::Componentize(
+    const TokenVector &tokens, const vector<string> &params,
+    ErrorHandler *herr) {
+  vector<Macro::FuncComponent> res;
+  std::unordered_map<std::string_view, size_t> params_by_name;
+  for (size_t i = 0; i < params.size(); ++i) params_by_name[params[i]] = i;
+  size_t e = 0;
+  for (size_t i = 0; i < tokens.size(); ++i) {
+    // -------------------------------------------------------------------------
+    // -- Handle concatenation markers. ----------------------------------------
+    // -------------------------------------------------------------------------
+    if (tokens[i].type == TTM_CONCAT) {
+      if (!i) {
+        herr->Error(tokens[i]) << "Concatenation marker (`##`) "
+            "cannot appear at either end of a replacement list.";
+        continue;
+      }
+      // Don't push spans CONCAT would always discard.
+      if (meaningful_span(tokens, e, i))
+        res.push_back(FuncComponent::TokenSpan{e, i});
+      res.push_back(FuncComponent::Paste{});
+      while (++i < tokens.size() && tokens[i].PreprocessesAway());
```

Comment:
i thought this type of syntax caused a warning in gcc? plus isn't it less confusing for readers to just put the continue?